eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpCommand.java

```diff
@@ -55,11 +57,13 @@ void warp(@Context Player player, @Context User user) {
     @DescriptionDocs(description = "Teleport to warp, if player has permission eternalcore.warp.bypass teleport will be instant", arguments = "<warp>")
     void warp(@Context Player player, @Arg Warp warp) {
         if (player.hasPermission("eternalcore.warp.bypass")) {
-            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), warp.getPosition(), Duration.ZERO);
-            return;
+            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), PositionAdapter.convert(warp.location()), Duration.ZERO);
         }
-
-        this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), warp.getPosition(), Duration.ofSeconds(5));
+        else {
+            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), PositionAdapter.convert(warp.location()), Duration.ofSeconds(5));
+        }
+        WarpEvent event = new WarpEvent(player,warp);
+        Bukkit.getPluginManager().callEvent(event);
```

Comment:
Use `EventCaller` to publish a warp teleport event.
Move this code to the top.

Check `event.isCancelled()` and if it is true, then return.

In the future, we can add post-event based on the CompletableFuture returned by the `TeleportTaskService#createTeleport`